### PR TITLE
Replace statefeed Initialize

### PR DESCRIPTION
**Pull request by kasey**
**What type of PR is this?**

Other

**What does this PR do? Why is it needed?**

This PR replaces the code that watches for the statefeed Initialize event with a simpler, race-free synchronization primitive based around blocking callers in a context-aware select until a sentinel channel is closed. The channel is only closed by a setter method that popualates the genesis data everyone is waiting on, timestamp and validator root. 

The returned type is also a step towards decoupling services from the blockchain service in order to determine the current slot. Instead, the returned type provides a Clock that can give the current time (Now) or current slot (CurrentSlot - which will allow it to satisfy existing interfaces).

One benefit of coupling these things together is that it allows code that uses the wall clock to begin using a synthetic clock instead, giving us a way to inject a global view on the passage of time that can move faster than 1:1 with the wall clock. This could be helpful going forward in building tests that move somewhat faster than realtime, as long as we aren't gated on other performance factors.

### CODE REVIEW

File: beacon-chain/slasher/service.go [hunk 1]
```diff
@@ -39,6 +39,8 @@ type ServiceConfig struct {
 	SlashingPoolInserter    slashings.PoolInserter
 	HeadStateFetcher        blockchain.HeadFetcher
 	SyncChecker             sync.Checker
+	ClockWaiter             startup.ClockWaiter
+	ClockSetter             startup.ClockSetter
```
Comment (nisdas) [inline beacon-chain/slasher/service.go:43]: why both ? slasher should only need a waiter

File: beacon-chain/startup/genesis.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+// GenesisValidatorsRoot returns the genesis state validator root
+func (g *Clock) GenesisValidatorsRoot() []byte {
+	return g.vr
+}
+
+// CurrentSlot returns the current slot relative to the time.Time value that Clock embeds.
+func (g *Clock) CurrentSlot() types.Slot {
+	return slots.Duration(g.t, g.now())
```
Comment (nisdas) [inline beacon-chain/startup/genesis.go:40]: same thing here

File: beacon-chain/startup/genesis.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+// WithNower allows tests in particular to inject an alternate implementation of time.Now (vs using system time)
+func WithNower(n Nower) ClockOpt {
+	return func(g *Clock) {
+		g.now = n
+	}
+}
+
+// NewClock constructs a genesis value, providing the ability to override the
+func NewClock(t time.Time, vr []byte, opts ...ClockOpt) *Clock {
```
Comment (nisdas) [inline beacon-chain/startup/genesis.go:60]: Where ever this constructor has been used, a `nower` has not been provided as an option. So the end result is that methods such as `CurrentSlot` and `Now` will panic. This should be fixed

File: beacon-chain/slasher/service.go (hunk 1 above)
Comment (kasey) [reply beacon-chain/slasher/service.go:43]: Thanks for the catch, I *think* it was previously needed for a test but after refreshing from develop I can't find that code, removing.

File: beacon-chain/startup/synchronizer.go
Comment (rkapka) [inline beacon-chain/startup/synchronizer.go:1]: Missing documentation in this file

File: beacon-chain/blockchain/service_test.go [hunk 4]
```diff
@@ -562,3 +496,19 @@ func TestChainService_EverythingOptimistic(t *testing.T) {
...
+// MockClockSetter satisfies the ClockSetter interface for testing the conditions where blockchain.Service should
+// call SetGenesis.
+type MockClockSetter struct {
+	G   *startup.Clock
+	Err error
+}
+
+var _ startup.ClockSetter = &MockClockSetter{}
+
+// SetGenesis satisfies the ClockSetter interface.
```
Comment (rkapka) [inline beacon-chain/blockchain/service_test.go:509]:
```suggestion
// SetClock satisfies the ClockSetter interface.
```

File: beacon-chain/sync/service.go [hunk 5]
```diff
@@ -164,14 +167,14 @@ func NewService(ctx context.Context, opts ...Option) *Service {
...
-	go r.registerHandlers()
-	go r.verifierRoutine()
-
 	return r
 }
 
 // Start the regular sync service.
 func (s *Service) Start() {
+	go s.verifierRoutine()
+	go s.registerHandlers()
```
Comment (rkapka) [inline beacon-chain/sync/service.go:176]: Why did you move this from `NewService`?

File: time/slots/slottime.go [hunk 6]
```diff
@@ -173,6 +173,14 @@ func CurrentSlot(genesisTimeSec uint64) primitives.Slot {
 	return primitives.Slot((now - genesisTimeSec) / params.BeaconConfig().SecondsPerSlot)
 }
 
+// Duration computes the slot of end relative to start.
```
Comment (rkapka) [inline time/slots/slottime.go:176]: How about
> Duration computes the number of full slots passed between from start until end

File: beacon-chain/sync/service.go (hunk 5 above)
Comment (kasey) [reply beacon-chain/sync/service.go:176]: Thanks for noticing this, it's something I want input on. I think the intent of the `NewService` pattern is to centralize initialization of values, so that service dependencies can be wired up. Whereas the intent of `Start` is to actually kick off the Service goroutines and runloops, after services have been configured and initialized. `sync.Service` is unique in actually starting up goroutines to do meaningful work in `NewService`, which muddles the distinction. I assumed that starting these goroutines in NewService was a hack to deal with raciness and make sure sync was able to register well ahead of the start event being triggered, which is no longer an issue. @nisdas do you have any thoughts?

**Review by rkapka — APPROVED** [review]
Great stuff

**Review by terencechain — COMMENTED** [review]
High level comments are both `beacon-chain/startup/clock.go` and `beacon-chain/startup/synchronizer.go` lacking unit tests. Do we want local unit tests in the `startup` pkg?

File: time/slots/slottime.go [hunk 7]
```diff
@@ -173,6 +173,14 @@ func CurrentSlot(genesisTimeSec uint64) primitives.Slot {
 	return primitives.Slot((now - genesisTimeSec) / params.BeaconConfig().SecondsPerSlot)
 }
 
+// Duration computes the span of time between two instants, represented as Slots.
+func Duration(start, end time.Time) primitives.Slot {
```
[the remark this answers is not in the document]
Comment (kasey) [reply time/slots/slottime.go:177]: added coverage for this

**Comment by kasey** [comment]
> High level comments are both `beacon-chain/startup/clock.go` and `beacon-chain/startup/synchronizer.go` lacking unit tests. Do we want local unit tests in the `startup` pkg?

@terencechain I added some test coverage for the startup package, please take another look.

**Review by terencechain — APPROVED** [review]
No further comments from me. Would be good to someone take a look at `p2p` and `sync` pkgs

File: beacon-chain/sync/service.go (hunk 5 above)
Comment (nisdas) [reply beacon-chain/sync/service.go:176]: I think in rare cases the initial sync service can send the `synced` event before the `sync` service starts up, if that happens the node will be stuck and unable to subscribe to gossip. To play on the safer side, the spawning off of the goroutines should be in the constructor for now. We can do something similar for the `synced` event in a follow up PR

**Comment by kasey** [comment]
@nisdas please look at commit https://github.com/prysmaticlabs/prysm/pull/12285/commits/8551fd12523cefa2d2965d2361a2b1793eb0869b

This would sidestep the initial-sync/sync race concern by removing the Synced event entirely. I'm hoping this doesn't push the PR over the edge any more than the other changes, but if it does I can remove this commit. I wanted to do it this way rather than a follow-up PR because I know a bunch of tests will go wonky if I move the startup goroutines from Start back to NewService, so I think this gets us to the finish line a little faster and hopefully actually reduces risk by getting rid of raciness.

**Comment by nisdas** [comment]
As a last step you should remove the `statefeed.Synced` and `statefeed.Initialized` events from the statefeed package @kasey so that we aren't using it anywhere unintentionally

**Review by nisdas — DISMISSED** [review]

**Review by nisdas — APPROVED** [review]
